Makefile change to support newer arm compilers (#1430)

prismv opened the pull request:
See also #1427 and #1066.

raprism committed:
proposed specs change #1430 (derived from #1427)

raprism committed:
FancyExponent used PRId64 format

Thereby snprintf is used. But PRId64 is not supported by newlib nano.
Change uses PRId32 for snprintf, and %ld for renderer r.printf

raprism committed:
Forgot cast :(

LdBeth commented:
To support hardware float printing, `-u _printf_float` needs to be added after `-specs=nano.specs`

prismv committed:
Merge branch 'c3d:stable' into ra-dev

raprism committed:
add option to support HWFP printf

acc. to https://github.com/c3d/db48x/issues/1430#issuecomment-2782014310

c3d committed:
decimal: Fix `sqrt` for squares

The terminating condition for sized `decimal::compare` was incorrect,
since it caused us to skip the trailing base-1000 digits in the
number. This caused `sqrt` to incorrectly terminate early, and user
@LdBeth remarked that this happens in particular for perfect squares
like `25` or `36`.

Reported-by: User @LdBeth
Signed-off-by: Christophe de Dinechin <christophe@dinechin.org>

raprism committed:
Merge remote-tracking branch 'origin/dev' into ra-dev

c3d commented on Makefile:
```diff
@@ -404,12 +404,13 @@ SIZE = arm-none-eabi-size
 BIN = $(OBJCOPY) -O binary -S
 CPUFLAGS += -mthumb -march=armv7e-m -mfloat-abi=hard -mfpu=fpv4-sp-d16
-PLATFORM_FLAGS = -Wno-packed-bitfield-compat -Wall -fdata-sections -ffunction-sections
+PLATFORM_FLAGS = -Wno-packed-bitfield-compat -Wall -fdata-sections -ffunction-sections -specs=nano.specs -specs=nosys.specs
```
Size comparison on my compiler gcc 10.3.1:

Your branch:
```
1097984	  93952	   3436	1195372	 123d6c	build/dm32/release/db50x.elf
 918768	   3656	   3432	 925856	  e20a0	build/dm42/release/db48x.elf
  907620 db50x.pg5
  642044 db48x.pgm
```

Original build:
```
1120240	  95184	   3444	1218868	 129934	build/dm32/release/db50x.elf
  931148 db50x.pg5
 940712	   4640	   3440	 948792	  e7a38	build/dm42/release/db48x.elf
  664980 db48x.pgm
```

So the new specs are indeed an improvement, even with the old toolchain.

c3d commented on src/decimal.cc:
```diff
@@ -660,7 +660,7 @@ RENDER_BODY(decimal)
             {
                 char expbuf[32];
-                size_t written = snprintf(expbuf, 32, "%" PRId64, dispexp);
+                size_t written = snprintf(expbuf, 32, "%" PRId32, dispexp);
```
This is a loss in capability for what I consider a very bad reason. Rendering a 64-bit value to a buffer is not that complicated. So please write a small helper function that renders a 64-bit value as decimal and use it where `snprintf` was used.

c3d reviewed (changes requested):
I see no reason to lose the ability to display 64-bit exponents.
@raprism, @LdBeth can you test my reworking of this PR here and check it works for you? 

https://github.com/c3d/db48x/compare/dev...c3d:db48x:ra-dev?expand=1

prismv commented:
Compromise is bare pointer arithmetics to substitute snprintf 25cd85e69275c3b8f6f5a9892ac931720d530157 by @c3d, so branch with last commit 1a87c2c seems to be a proper solution.
